feat: generate types for BigQueryDataTransferConfig

jingyih opened the pull request.

jingyih commented:
Hi @justinsb, it seems that the current approach of visiting proto files is leading to two issues:
 - We are generating some unnecessary Go structs, such as `TransferRun`.
 - We are missing the generation of certain dependent Go structs, like `google_protobuf_Struct`.
 
Should we consider switching to "find and generate the main resource proto message and its dependent messages" approach?

jasonvigil commented:
> Hi @justinsb, it seems that the current approach of visiting proto files is leading to two issues:
> 
> * We are generating some unnecessary Go structs, such as `TransferRun`.
> * We are missing the generation of certain dependent Go structs, like `google_protobuf_Struct`.
> 
> Should we consider switching to "find and generate the main resource proto message and its dependent messages" approach?

+1, I am seeing extra resources generated here also: https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2579

jingyih commented on config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigquerydatatransferconfigs.bigquerydatatransfer.cnrm.cloud.google.com.yaml:
```diff
@@ -123,128 +139,81 @@ spec:
-                  be created with this service account credentials. It requires that
-                  requesting user calling this API has permissions to act as this service account.
-                type: string
+              userID:
```
Should we hide this field?

jingyih commented:
/assign @justinsb 

Hi Justin, I made some changes to the type generator in this PR. Could you help take a look?

jingyih commented on config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigquerydatatransferconfigs.bigquerydatatransfer.cnrm.cloud.google.com.yaml:
```diff
@@ -123,128 +146,87 @@ spec:
+                      transfer can be triggered manually is not limited by this option.
                     type: string
                 type: object
-              sensitiveParams:
```
This field only exists in Terraform.
https://registry.terraform.io/providers/hashicorp/google-beta/latest/docs/resources/bigquery_data_transfer_config#sensitive_params

jingyih commented on config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigquerydatatransferconfigs.bigquerydatatransfer.cnrm.cloud.google.com.yaml:
```diff
@@ -123,128 +146,87 @@ spec:
-                required:
-                - secretAccessKey
-                type: object
-              serviceAccountName:
```
This is also a terraform only field.
https://registry.terraform.io/providers/hashicorp/google-beta/latest/docs/resources/bigquery_data_transfer_config#service_account_name

yuwenma commented on apis/bigquerydatatransfer/v1alpha1/bigquerydatatransferconfig_types.go:
```diff
@@ -0,0 +1,197 @@
+	//  https://cloud.google.com/bigquery-transfer/docs/cloud-storage-transfer#bq
+	Params map[string]string `json:"params,omitempty"`
+
+	// +required
+	Parent `json:",inline"`
```
Thank you for using this style!  It looks nicer in code than in the guide 😆

jingyih replied:
I found it in your doc! Maybe we should add `Parent` to the controller builder template?

yuwenma commented on apis/bigquerydatatransfer/v1alpha1/bigquerydatatransferconfig_types.go:
```diff
@@ -0,0 +1,197 @@
+	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+
+	// A unique specifier for the BigQueryDataTransferConfig resource in GCP.
+	// +optional
```
super nit: we can ignore the `+optional` since this is the default and the status is not a user specified field.

jingyih replied:
These come from the controller builder template. Removing them in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2613.

yuwenma commented on config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigquerydatatransferconfigs.bigquerydatatransfer.cnrm.cloud.google.com.yaml:
```diff
@@ -6,20 +6,16 @@ metadata:
     cnrm.cloud.google.com/managed-by-kcc: "true"
-    cnrm.cloud.google.com/stability-level: alpha
     cnrm.cloud.google.com/system: "true"
-    cnrm.cloud.google.com/tf2crd: "true"
```
This means we need to provide the direct controller, but I don't see it in this PR. Since this is a alpha resource, could you add 
 `labels="cnrm.cloud.google.com/tf2crd=true"` in the types.go? Example  
https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/660ef0bdad70d384e0c742a4222d97743c582aaf/apis/kms/v1beta1/kmskeyring.go#L73

I'll update the guide to capture this.

yuwenma replied:
Actually, you can add this resource to the `excluded_resources` to not override the CRD until the controller is ready (especially since this resource has API changes). https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/dev/tasks/generate-crds#L58

yuwenma reviewed (commented):
/lgtm
/approve

/hold 
Looks really good! One comment about continue using the TF_based controller until the direct controller is ready (otherwise this alpha resource will be broken).

jingyih commented on apis/bigquerydatatransfer/v1alpha1/bigquerydatatransferconfig_types.go:
```diff
@@ -0,0 +1,197 @@
+	//
+	//  The format for specifying a pubsub topic is:
+	//  `projects/{project_id}/topics/{topic_id}`
+	NotificationPubsubTopic *string `json:"notificationPubsubTopic,omitempty"`
```
Thanks! Given that is resource is alpha, I was thinking about adding reference in a separate PR, but I don't mind add it in this PR. I will update this PR.

jingyih commented on apis/bigquerydatatransfer/v1alpha1/types.generated.go:
```diff
@@ -0,0 +1,55 @@
+package v1alpha1
+
+// +kcc:proto=google.cloud.bigquery.datatransfer.v1.EmailPreferences
+type EmailPreferences struct {
```
Good catch. This is indeed a weird name.

https://github.com/googleapis/googleapis/blob/9c6ceea874182fbfda3e59faba5df35906567c5b/google/cloud/bigquery/datatransfer/v1/transfer.proto#L72-L77

jingyih commented:
Rebased to upstream/master.

Added 3 new commits.
 - [cdbc922](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2573/commits/cdbc922f0078ea9067af42c90c2679c47009b4a1): some minor fix in types
 - [4d496d3](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2573/commits/4d496d3539c4bd8146111ae120761718777e0857): reference fields
 - [eed4dc9](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2573/commits/eed4dc9a2144c888d49f06e0d4239c1198e7075d): do not switch to direct controller yet

yuwenma commented:
/lgtm
/hold cancel

PR looks great! I think it's ready to go in once the broken presubmit passes.

jingyih commented:
Rebased to pick up the sample test fix in HEAD.

yuwenma commented:
/lgtm

jingyih commented:
```
--- FAIL: TestAllInSeries/samples/workerpool-with-peered-network
```

Retrying
